consolidation/consolidate.go

```diff
@@ -115,13 +130,48 @@ func nudge(start, preAggInterval, aggNum uint32) (uint32, int) {
 	postAggInterval := preAggInterval * aggNum
 	var num int
 	var diff uint32
+	// two important principles here:
+	// 1) aggregation buckets have timestamps that are cleanly divisible by postAggInterval.
+	// 2) we never make data pretend to be able to predict the future,
+	//    e.g. when aggregating a spike, the spike should never move to an earlier timestamp
+	//    in other words, data should only ever move to a later timestamp when being aggregated, never to the past.
+	// it follows that the start of (first point going into) an aggregation bucket
+	// is the point that has a timestamp of 1 preAggInterval after the cleanly-divisible-by-postAggInterval timestamp.
+
+	// let's clarify all this with an example.
+	// consider: preAggInterval = 5, aggNum = 4 => postAggInterval = 20
+	// points marked with a * are the first points going into an aggregation interval
+
+	// input   consolidate into    output points
+	// 15      ▼
+	// 20        ►                 20
+	// 25 *    ▼
+	// 30      ▼
+	// 35      ▼
+	// 40        ►                 40
+	// 45 *    ▼
+	// 50      ▼
+	// 55      ▼
+	// 60        ►                 60
+	// 65 *    ▼
+	// 70      ▼
+	// 75      ▼
+	//                             80
+
+	// with our nudging approach, we strip all the points that do not form a full aggregation bucket.
+	// IOW the desired first point must be the the first point of a bucket.
+	// in the example above, it means we strip the first 2 points and start at ts=25 (which will go into ts=40).
+
```

Comment:
it is.  it's also how we've implemented runtime consolidation in MT since forever
it's based on a guiding principle that i came up with but that is afaik in the wider graphite ecosystem not often really discussed or cared about and as a result also implemented oppositely across different tools (eg statsd also postmarks like MT). see https://grafana.com/blog/2016/03/03/25-graphite-grafana-and-statsd-gotchas/#timestamps for details
basically the idea is that data should never move into the past, but that people can expect data to be delayed (and move into the future instead). e.g. visually, a spike should never appear on a chart at an earlier time then when it actually happened, and people should expect that with consolidation, data will move it a bit to the right. so that whenever you look at a point in time (whether aggregated or not) you can always know that happened at some point in time since the previous timestamp.

perhaps this is just one of my quixotic quirks? is there a flaw in my reasoning?
especially now that we care more about stricter compatibility with graphite, do we want to undo this and do things the graphite way? though i would rather keep the postmarking, i'm open to discussion and/or change.